### ISSUE TITLE
Interpreter binaries not found in the current virtual environnement

### ISSUE THREAD

**Issue by rsuhada** [jorgenschaefer/elpy#1370]
# Summary

When running working on my virtualen by `M-x pyven-workon tensorflow` I get an error:

`elpy-shell-get-or-create-process: Interpreter binaries not found in the current virtual environnement` and the inferior python shell does not start.

When I run the same environment in terminal (outside of Emacs) everything works fine.

Thank you very much!

# My configuration
## OS
Ubuntu 16.04
GNU Emacs 25.1.2

## Result of `(elpy-config)`
```
Virtualenv........: tensorflow (/home/rxs/.virtualenvs/tensorflow)
RPC Python........: 3.5.2 (/home/rxs/.virtualenvs/tensorflow/bin/python)
Interactive Python: ipython3 (/usr/local/bin/ipython3)
Emacs.............: 25.1.2
Elpy..............: 1.20.0
Jedi..............: 0.11.1 (0.12.0 available)
Rope..............: Not found (0.10.7 available)
Autopep8..........: 1.3.3 (1.3.5 available)
Yapf..............: 0.20.1 (0.21.0 available)
Syntax checker....: flake8 (/usr/local/bin/flake8)
```

## Elpy configuration in my init.el
```
(use-package elpy
  :ensure t
  :init
  (add-to-list 'warning-suppress-types '(undo discard-info))
  :config
  (elpy-enable)
  (flymake-mode-off)
  (setq elpy-rpc-backend "jedi")
  (setq elpy-rpc-python-command "python")
  (defun elpy-goto-definition-or-rgrep ()
    "Go to the definition of the symbol at point, if found. Otherwise, run `elpy-rgrep-symbol'."
    (interactive)
    (ring-insert find-tag-marker-ring (point-marker))
    (condition-case nil (elpy-goto-definition)
      (error (elpy-rgrep-symbol
              (concat "\\(def\\|class\\)\s" (thing-at-point 'symbol) "(")))))
  (define-key elpy-mode-map (kbd "M-.") 'elpy-goto-definition-or-rgrep)
  (setq
   python-shell-interpreter "ipython3"
   python-shell-interpreter-args "--simple-prompt --pprint")
```

**Comment by galaunay** [comment]
Activating a virtualenv basically just add its binaries folder to the PATH.
It means that if a binary (let's say `ipython3`) is not present on the virtualenv, the system binary (`/usr/local/bin/ipython3` for example) will be used in place.

This warning was added to prevent this situation.
It just tell you that the interpreter you asked (`ipython3`) is not available on the virtualenv you have activated.

Is it possible that you forgot to install ipython3 in your virtualenv ?
(else, there is a bug in Elpy...).

**Comment by rsuhada** [comment]
Thank you. It is true that I did not install `ipython3` in the virtual environment, since it is available on the system. 

If I activate my virtualenv on the terminal, and run `iptyhon3`, everything works fine (so virtualenv packages are available from within ipython). In fact, if I try `pip install ipython` when the env is active I get the message that the requirement is already satisfied...

If Elpy just gave a warning, it would be ok, but no inferior python shell is started up... (I updated this in the original questions).

Just so that I understand - Is really my only option to somehow *force* `ipython` installation from within virtualenv even if pip says nothing is needed?

**Comment by galaunay** [comment]
> Thank you. It is true that I did not install ipython3 in the virtual environment, since it is available on the system.
>
>If I activate my virtualenv on the terminal, and run iptyhon3, everything works fine (so virtualenv packages are available from within ipython).

I will make a PR to make it raise a warning instead.
You can put the following bit of elisp in your `init.el` after elpy initialisation, as a temporary workaround:
```Elisp
(defun elpy-shell-get-or-create-process (&optional sit)
  "Get or create an inferior Python process for current buffer and return it.

If SIT is non-nil, sit for that many seconds after creating a
Python process. This allows the process to start up."
  (let* ((bufname (format "*%s*" (python-shell-get-process-name nil)))
         (proc (get-buffer-process bufname)))
    (if proc
        proc
      (let ((default-directory (or (and elpy-shell-use-project-root
                                        (elpy-project-root))
                                   default-directory)))
        (run-python (python-shell-parse-command) nil t))
      (when sit (sit-for sit))
      (when (elpy-project-root)
        (python-shell-send-string
         (format "import sys;sys.path.append('%s')" (elpy-project-root))))
      (get-buffer-process bufname))))
```


> In fact, if I try pip install ipython when the env is active I get the message that the requirement is already satisfied...
> 
> Just so that I understand - Is really my only option to somehow force ipython installation from within virtualenv even if pip says nothing is needed?

This is weird, are you sure pip is available in the virtualenv ?
what is the result of `where pip` and  `where ipython` ?

**Comment by rsuhada** [comment]
Thank you.

I activate the environment: `workon tensorflow`

`where pip` gives:
```
~/tensorflow/bin/pip
/usr/local/bin/pip
```
`where ipython`:
```
/usr/local/bin/ipython
```
If I do `pip install ipython`:
`Requirement already satisfied: ipython in /usr/local/lib/python3.5/dist-packages (6.2.1)`

**Comment by galaunay** [comment]
Thanks for the details, 
I didn't know non-isolated virtualenv behaved this way.

This issue should be fixed as soon as #1372 is merged.

Please re-open this if it is not the case.
